llvm/lib/Target/EVM/EVMConstantSpiller.cpp

```diff
@@ -0,0 +1,100 @@
+//===----- EVMConstantSpiller.cpp - Spill constants to memory --*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file identifies IMM_RELOAD instructions representing spilled constants
+// throughout the module. It spills constants at the start of the entry function
+// and replaces IMM_RELOAD with the corresponding reload instructions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "EVMConstantSpiller.h"
+#include "EVMInstrInfo.h"
+#include "EVMSubtarget.h"
+#include "MCTargetDesc/EVMMCTargetDesc.h"
+#include "TargetInfo/EVMTargetInfo.h"
+#include "llvm/CodeGen/MachineFrameInfo.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "evm-spill-constants"
+
+constexpr uint64_t SpillSlotSize = 32;
+
+static MachineInstr *emitPush(MachineBasicBlock &MBB,
+                              MachineBasicBlock::iterator InsertBefore,
+                              const EVMInstrInfo *TII, const APInt &Imm,
+                              LLVMContext &Ctx, const DebugLoc &DL) {
+  unsigned Opc = EVM::getPUSHOpcode(Imm);
+  auto MI = BuildMI(MBB, InsertBefore, DL, TII->get(EVM::getStackOpcode(Opc)));
+  if (Opc != EVM::PUSH0)
+    MI.addCImm(ConstantInt::get(Ctx, Imm));
+  return MI;
+}
+
+uint64_t EVMConstantSpiller::getSpillSize() const {
+  return ConstantToUseCount.size() * SpillSlotSize;
+}
+
+void EVMConstantSpiller::emitSpills(uint64_t SpillOffset,
+                                    MachineFunction &EntryMF) {
+  LLVMContext &Ctx = EntryMF.getFunction().getContext();
+  const EVMInstrInfo *TII = EntryMF.getSubtarget<EVMSubtarget>().getInstrInfo();
+
+  DenseMap<APInt, uint64_t> ConstantToSpillOffset;
```

Comment:
Maintainability: Spill offset == Num of const in ConstantToUseCount map * 32. Do we need a separate map for it?

Reply:
It's convenient to have a map with final offsets because, as we refer to this info twice. The second time if when we walk over reloads.